text/1028-cli-hooks.rst

```diff
@@ -56,70 +69,184 @@ are always executed in WSL.
--------------------
+Command Hooks
+-------------
 
 This category of hooks are intended for one-time setup of important fixtures
-or application configuration. They are perfect for setting up data fixtures or
-exporting connection strings for non-gel tools to use. There are generally two
-scenarios when this kind of hook might be needed:
-
-1) After ``gel project init`` the ``project.init.after`` hook will be
-   executed. This is good for updating any additional configuration that
-   depends on the project database. It is also a good place for data fixture
-   scripts.
-
-2) After ``gel branch wipe`` the ``branch.wipe.after`` hook will be executed.
-   This may be a good place for restoring the data fixtures. However it is
-   probably unnecessary to run any configuration scripts at this time. This is
-   why this hook does not simply run all the same scripts as the
-   ``project.init.after``.
-
-3) After ``gel branch switch`` command that changes the current branch the
-   ``branch.switch.after`` hook will be executed. This is a good place for
-   updating any configuration scripts (e.g. updating Postgres connection
-   string). This hook can also be used to keep the source code branch in sync
-   with the database branch. This is probably not a good hook for running
-   data fixtures as the data in branches in unaltered between switches.
-
-If ``project init`` runs migrations, the ``project.init.after`` hook is
-triggered *before* the ``migration.apply.after`` hook. The motivation is that
-whatever is necessary as a one-time project setup is likely to be a
-pre-requisite for migration hooks.
-
-Similarly, ``branch.wipe.after`` and ``branch.switch.after`` hooks are
-executed before the migration hook triggers.
-
-
-Migration Hooks
----------------
-
-This category of hooks is intended for scripts that sync up the codebase with
-the schema state. They are defined by the ``migration.apply.after`` setting.
-It may be re-compiling the query-builder or reflected individual queries. It
-may be generating an ORM compatibility layer. It could also perform any
-project-specific custom operation.
+or application configuration. They are perfect for setting exporting
+connection strings for non-gel tools to use. They are primarily triggered by
+the corresponding CLI commands, but sometimes commands may trigger more than
+one hook:
+
+* ``gel project init`` command triggers the ``project.init.before``
+  and ``project.init.after`` hook. If the migrations are applied at the end of
+  the initialization, then the ``migration.apply.before`` and
+  ``migration.apply.before`` hooks are also triggered.
+* ``gel branch switch`` command triggers ``branch.switch.before``,
+  ``schema.update.before``, ``branch.switch.after``, and ``schema.update.after``
+  hooks in that relative order.
+* ``gel branch wipe`` command triggers the ``branch.wipe.before``,
+  ``schema.update.before``, ``branch.wipe.after``, and ``schema.update.after``
+  hooks in that relative order.
```

Comment:
`edgedb branch wipe` doesn't wipe the schema, only date, does it?

Reply:
`branch.wipe.before` might abort the operation.

BTW, @aljazerzen, can we set some environment variables for the hook commands we run? Like for `branch.wipe.before` we can set 

```
env GEL_HOOK_NAME=branch.wipe.before \
    GEL_HOOK_INSTANCE=1st1/production
    GEL_HOOK_BRANCH=main
```

and then people will be able to write scripts that protect/react to things.

We don't need to implement this now, but I'd like the RFC to have a "future extensions" section that mentions this. And we should discuss what happens if a "before" hook errors out (IMO the command needs to be aborted), and what happens if an "after" hook errors out (IMO there's not much we can do, so just face plant).

Reply:
> `edgedb branch wipe` doesn't wipe the schema, only data, does it?

It actually does `reset schema to initial;` which does wipe the schema itself along with the data. FWIW, it would be awesome if there was a data-only version of this command since manually "clearing" a branch is actually a bit annoying.

Reply:
It does wipe the schema, not just data.